### PR TITLE
Introduce nomopyomo

### PR DESCRIPTION
**Pull request by FabianHofmann**
Introduce optimization without usage of pyomo, based on [nomopyomo](https://github.com/PyPSA/nomopyomo/blob/master/nomopyomo.py) which is way faster and memory efficient for especially large networks. 

What happened so far:

## New Modules

- linopf.py -  includes all code for setting up the problem
- linopt.py - includes all solver (io to solver) relevant code
- stats.py - include code for post-optimization analysis
 

## Ordinary features of lopf without pyomo:

- [x] include ramp limits
- [x] include Storage Units
- [x] calculate nodal imbalances
- [x] constant term in objective function
- [x] calculate voltage angles
- [x] implement glpk solver
- [x] logfile for cbc

## Additional features

- [x] global constraints for transmission expansion volume/cost limit
- [x] ability to fix values via '_set' for arbitrary variables, also static variables (so f.e. n.generators.p_nom_set can fix values for extendable p_nom variables)
- [x] warm start
- [x] add iterative lopf (ilopf), updating impedances after each solving process 
- [x] docstrings for `linop{t,f}.py`
- [x] add docstrings to documentation (+ incorporate changes)
- [x] ensure windows compatibility
- [x] add windows and osx in travis CI 

 
## Other changes 
 
- determine_network_topology now includes find_slackbus and  find_cycles

- add new lopf to tests in   
   1. [acdc test](https://github.com/PyPSA/PyPSA/blob/nomopyomo/test/test_ac_dc_lopf.py)
   2. [storage unit test](https://github.com/PyPSA/PyPSA/blob/nomopyomo/test/test_opf_storage.py)

- skip python2.7 test in travis file. I had to much python3 code syntax, and didn't see the nessecity to rewrite it as so many package give up on python2 support. But perhaps you have a different view.

The network.lopf function got a new argument as proposed in #96, so it becomes possible now to call
```
network.lopf(pyomo=False)
```

### CODE REVIEW

**Comment by fneum** [comment]
This already looks super nice! I just added "docstrings in `linop{t,f}.py` and documentation" to the remaining todos.

**Comment by fneum** [comment]
I'm fine with dropping Python 2.7 support in future versions. This discussion already came up in #79 . Obviously, we should make this clear in release notes and documentation. Additionally, we should keep a documentation version of the last Python 2.7 support version of PyPSA in pypsa.readthedocs.io.

**Comment by FabianHofmann** [comment]
> This already looks super nice! I just added "docstrings in `linop{t,f}.py` and documentation" to the remaining todos.

Thanks! And yes. Could you help me there? Perhaps with the documention part, always missing the patience to do it properly. Considering the docstrings, most of the function in linopt already have it already. 
Missing parts are 
- [ ] define_ functions in linopf.py (which probably should just get a small note)
- [ ] prepare_lopf and assign_solution in linopf.py
- [ ] run_and_read_ functions in linopt.py

**Comment by fneum** [comment]
> Could you help me there?
Sure, I can do that. WIP will be here: https://pypsa.readthedocs.io/en/nomopyomo/

**Comment by FabianHofmann** [comment]
So... the doc is now updated and includes most of the necessary information. It's not fully perfect but good enough :) All doc changes took place in [here](https://pypsa.readthedocs.io/en/nomopyomo/optimal_power_flow.html)  (nearly no deletions, only added sections) 

The module does not yet include
* unit commitment 
* extracting non-optimal solutions

which however can be postponed. The state is good enough to be merged now. @fneum can you perhaps double-check the documentation part? It would be nice to have an approval from @nworbmot if he agrees to merging. The new feature could make it then into the next version which necessary anyway due to #102.

File: doc/optimal_power_flow.rst [hunk 1]
```diff
@@ -213,7 +187,7 @@ availability is a constant.
 
 
 If the generator's nominal power :math:`\bar{g}_{n,s}` is also the
-subject of optimisation (``generator.p_nom_extendable == True``) then
+subject of optimisation (``generator.p_nom_extendable -- True``) then
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply doc/optimal_power_flow.rst]: aah, no, that was my editor when I changed underlining of the headers, too optimistic it wouldn't affect any other things

File: environment.yaml [hunk 2]
```diff
@@ -14,3 +14,4 @@ dependencies:
   - cartopy>=0.16
   - coincbc
   - glpk
+  - gurobi::gurobi
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply environment.yaml]: gurobipy is needed to solve the lp problem without pyomo and as for as I know, does not come automatically with pyomo (which is also in the dependency)

File: pypsa/pf.py [hunk 3]
```diff
@@ -833,7 +833,7 @@ def find_tree(sub_network, weight='x_pu'):
 
     #find bus with highest degree to use as slack
     tree_slack_bus, slack_degree = max(degree(sub_network.tree), key=itemgetter(1))
-    logger.info("Tree slack bus is %s with degree %d.", tree_slack_bus, slack_degree)
+#    logger.info("Tree slack bus is %s with degree %d.", tree_slack_bus, slack_degree)
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply pypsa/pf.py]: the logging comes up multiple times during the lopf and distracts a bit from the important logging. I found it a bit fussy, when it tells you that often, which are the slackbuses now :) But I can definitely include it again.

**Comment by FabianHofmann** [comment]
> Check on documentation. Looks great, only few minor things.

thanks @fneum for looking at it, I will include the other things later

File: pypsa/opt.py [hunk 4]
```diff
@@ -382,3 +386,11 @@ def wrapper():
...
         logger.debug("Unable to measure memory usage, since the resource library is missing")
         return False
+
+
+# =============================================================================
+# Helpers for opf_lowmemory
+# =============================================================================
+
+
+
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply pypsa/opt.py]: yes

File: environment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (neunzehnachtneun) [reply environment.yaml]: Jep, that's true. gurobi doesn't come with pyomo. But as I understand the script it's possible to chose between cbc and gurobi, isn't it?
I would suggest to remove gurobi as a dependency, as it's a commercial solver and the users of pypsa should decide on their own to buy and install gurobi or not.

[the remark this answers is not in the document]
Comment (FabianHofmann) [reply environment.yaml]: Yes it is, supported for now are 
* glpk
* cbc
* gurobi

I totally agree on the point of letting the user decide which solver to use. But is does not contradict. Installing gurobipy means only installing the interface, not the solver itself. This counts for the other solvers as well, none of them is installed pypsa. Like this the code can guarantee, that as soon as the solver is installed, the optimization runs.

[the remark this answers is not in the document]
Comment (FabianHofmann) [reply environment.yaml]: but if you guys don't agree, I don't mind taking it out :)

File: pypsa/pf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (coroa) [reply pypsa/pf.py]: Set it as `logger.debug`?

[the remark this answers is not in the document]
Comment (FabianHofmann) [reply pypsa/pf.py]: yes! that is way better

File: pypsa/linopf.py [hunk 5]
```diff
@@ -720,6 +720,11 @@ def network_lopf(n, snapshots=None, solver_name="cbc",
     if formulation != "kirchhoff":
         raise NotImplementedError("Only the kirchhoff formulation is supported")
 
+    if n.generators.committable.any():
+        logger.warn("Unit commitment is not yet implemented for optimsation "
+           "without using pyomo. The following generators will be treated as "
+          f"non-commitables:\n{list(n.generators.query('committable').index)}")
```
Comment (coroa) [inline pypsa/linopf.py]: optimsation -> optim*i*sation
non-commitables -> non-commit*t*ables

Unit *commitment* is correct.

Comment (FabianHofmann) [reply pypsa/linopf.py]: yep

**Comment by FabianHofmann** [comment]
After talking to @nworbmot the feature 
> fix values via '_set' for arbitrary variables, also static variables (so f.e. n.generators.p_nom_set can fix values for extendable p_nom variables)

will preliminary only work for `state_of_charge_set` (similar to the normal lopf). However it can be easily extended to other variables (see commenting of the two lines in 8cfce1f05aaad9716df9aebc855139dae762bf10)

**Comment by nworbmot** [comment]
The nomopyomo branch is not working properly for sector coupling examples. I fixed the stores emissions bug, but it's still not giving the right answer. I think the problem is the missing implementation of links with multiple outputs. This works in regular PyPSA and the [original nomopyomo code](https://github.com/PyPSA/nomopyomo).
@FabianHofmann could you fix this? For testing, see [this multilink example](https://pypsa.org/examples/biomass-synthetic-fuels-carbon-management.html).

**Comment by FabianHofmann** [comment]
ah multilink, yes I'll check it. Is there a specific reason why the solving takes longer? The small memory might be there gap is there because of keeping the references

**Comment by nworbmot** [comment]
The solving time difference is down to the fact that it's a different problem without the multilinks, and to random number choices gurobi makes - nothing that the python problem-building can influence.

**Comment by FabianHofmann** [comment]
@nworbmot, alright this should work now. It seems that the solution is non-unique. nomopyomo gets exactly the same objective value and same amount of captured carbon but flow in links are a bit different...

**Comment by nworbmot** [comment]
Great! I just tested it on my PyPSA-Eur-Sec test model and the objective was within the tolerance. I think we're there! I'll do a few more tests, then we can merge it into the master. Memory and speed look good:
pyomo-versus-fabian-191108-2010

**Comment by nworbmot** [comment]
FYI I'm just testing a few more things, e.g. how easy it is to extend constraints with extra_functionality etc.
I think the code will probably break in Windows, because of the way the tmp files are currently handled (stored in /tmp) and concatenated. We should either have a warning that it doesn't work for Windows, or have a platform-independent implementation.

**Comment by coroa** [comment]
For concatenation you can use: `shutil.copyfileobj` from standard library.

**Comment by FabianHofmann** [comment]
> 1. Having the variables cCounter and vCounter stored in the global linopt library is a bad idea, because there may be multiple networks in scope being optimised separately but sharing these variables. These variables should be stored locally on the network object, like in the original nomopyomo, e.g. network._cCounter and network._vCounter.

Yes, that makes sense

> 2. It's not 100% clear how to customise the new structure for extra constraints. Could you construct an example like I did for the original nomopyomo [here](https://github.com/PyPSA/nomopyomo#customisation) using extra_functionality?

There is a rough explanation in the [docs](https://pypsa.readthedocs.io/en/nomopyomo/optimal_power_flow.html#pyomo-is-set-to-false). But i'll do another example

> 3. The LOPF doesn't seem to work with GLPK.

Can you say what the error is? The [tests in the continuous integration](https://github.com/PyPSA/PyPSA/blob/nomopyomo/test/test_opf_storage.py) check the pyomo=False execution with glpk.  

> 4. I think the code will probably break in Windows, because of the way the tmp files are currently handled (stored in /tmp) and concatenated. We should either have a warning that it doesn't work for Windows, or have a platform-independent implementation using "from pyutilib.services import TempfileManager".

I try to make it compatible with windows, taking @coroa 's advice

**Comment by nworbmot** [comment]
Thanks for pointing me towards docs, that's helpful.
For GLPK the error was at:
sol = pd.read_fwf(data).set_index('Row name')
"KeyError: "None of ['Row name'] are in the columns"
Using the 3-node LOPF example:
https://github.com/PyPSA/PyPSA/blob/master/examples/minimal_example_lopf.py
It could be a differen GLPK version...

**Comment by fneum** [comment]
The tests are failing because `coincbc` is not available for Windows via `conda`. I noticed recently when setting up the test pipeline in pypsa-eur. (cf. https://anaconda.org/conda-forge/coincbc).

`GLPK` is available for all platforms via `conda` however (cf. https://anaconda.org/conda-forge/glpk). Does anything speak against using `GLPK` in favor of `cbc`?

This would require removing `coincbc` in the `environment.yaml`.

**Comment by FabianHofmann** [comment]
Yes, you are right. The thing is that most all lopf-tests are checking both solvers, which is kind of good as nomopyomo has different `run_and_read function`s for cbc and glpk. On the other hand the implementation for gurobi is also not validated...

So, I'll drop the cbc checks for now. Perhaps for later, there is a nice solution excluding windows machine from cbc implementation checks

**Comment by fneum** [comment]
One could also figure out how to install cbc for windows via command line for the `.travis.yml`, and make it conditional depending on the operating system.

**Comment by FabianHofmann** [comment]
sorry for 'spamming' a lot a commits here. I had to figure out how to install cbc solver only for non-windows machines in the travis CI. But it works now. Tomorrow, I'll reintegrate cbc in the tests

File: .travis.yml [hunk 6]
```diff
@@ -1,32 +1,23 @@
...
 sudo: false # Use container-based infrastructure
 
+
 matrix:
   include:
-    - env:
-      - PYTHON_VERSION="2.7"
+    - os:
+      - windows
+      - linux
```
Comment (fneum) [inline .travis.yml]: Add
```
- osx
```
for completeness?

File: .travis.yml [hunk 7]
```diff
@@ -1,32 +1,23 @@
...
-  - export PATH="$HOME/miniconda/bin:$PATH"
-  - hash -r
-  - conda config --set always_yes yes --set changeps1 no
-  - conda update -q conda
-  # Useful for debugging any issues with conda
-  - conda info -a
-  - source $HOME/miniconda/etc/profile.d/conda.sh
+  - wget https://raw.githubusercontent.com/trichter/conda4travis/latest/conda4travis.sh -O conda4travis.sh
+  - source conda4travis.sh
+  - if [ "$TRAVIS_OS_NAME" != "windows" ]; then echo "  - coincbc" >> environment.yaml; fi;
```
Comment (fneum) [inline .travis.yml]: Very creative!

File: .travis.yml (hunk 6 above)
Comment (FabianHofmann) [reply .travis.yml]: yes, is in

File: .travis.yml (hunk 7 above)
Comment (FabianHofmann) [reply .travis.yml]: :) best solution for now. Conditional installation in `environment.yaml` is unfortunately not supported, and I did not want to set up a third environment file

**Comment by FabianHofmann** [comment]
@nworbmot 
The points seem to be solved now. Just for revising 
* My GLPK version is 4.65, the ones in the CI are also not raising a key error. Could you perhaps check the code with an updated GLPK version? 
* Windows is now completely covered. OSX and Windows now went into the travis CI

**Comment by nworbmot** [comment]
@FabianHofmann Thank you for all your hard work! Brilliant! I'm tied up today and tomorrow, but I've set aside Friday for testing and merging into the master branch. Then we can even do a new release, so people can test it in the wild.

**Comment by FabianHofmann** [comment]
@nworbmot  Great! thanks!

**Comment by nworbmot** [comment]
I'm still struggling a bit with customising the nomopyomo with extra_functionality. Two questions:
1. When should I be passing axes around from linexpr to write_constraint? If I don't do it, then it write constraints with the same name and gurobi complains. The documentation is a bit opaque: "If lower and upper are numpy.ndarrays it axes must not be None but a tuple of (index, columns) or (index)."
2. Suppose I want to make a custom constraint, e.g. a line volume constraint. How do I extract the shadow price?

**Comment by nworbmot** [comment]
I think we're also going to need documentation on how to define new variables.
Sorry to be so difficult, but I'm trying to anticipate how users might use the software. The new version of nomopyomo is elegant, but it's also more complicated.

**Comment by FabianHofmann** [comment]
> I'm still struggling a bit with customising the nomopyomo with extra_functionality. Two questions:
> 
> 1. When should I be passing axes around from linexpr to write_constraint? If I don't do it, then it writes constraints with the same name and gurobi complains. The documentation is a bit opaque: "If lower and upper are numpy.ndarrays it axes must not be None but a tuple of (index, columns) or (index)."
> 2. Suppose I want to make a custom constraint, e.g. a line volume constraint. How do I extract the shadow price?

The `write_constraint` needs an axes argument if all of the former arguments (`lhs`, `sense`, `rhs`) are pure numpy arrays. Then the return value is a pd.Series or DataFrame, which can be attached to the network via `set_conref`. 

I think in all cases when using the `linexpr` for custom constraints it is preferable to set `as_pandas` to `True`, so for the LV constraint it would be 

```
ext_i = n.lines.query('s_nom_extendable').index
# get all variables
terms = linexpr((n.lines.length[ext_i], get_var(n,  'Line', 's_nom')[ext_i]))
lhs += join_exprs(v) + '\n'

rhs = someconstant
con = write_constraint(n, lhs, '<=', rhs, axes=pd.Index([constraintname]))
set_conref(n, con, 'GlobalConstraint', 'mu', False, name)

```
You are right, it is a bit inelegant that one has to pass a pd.Index([name]) here. I think I can rewrite it in way if the length of the constraint is 1 that one can just simply pass a constraint name

**Comment by FabianHofmann** [comment]
> I think we're also going to need documentation on how to define new variables.
> Sorry to be so difficult, but I'm trying to anticipate how users might use the software. The new version of nomopyomo is elegant, but it's also more complicated.

No worry, it is good to make the first barrier easy for users

**Comment by nworbmot** [comment]
Thanks @FabianHofmann! What is
`lhs += join_exprs(v) + '\n'`

**Comment by FabianHofmann** [comment]
I just realized that in the example I didn't use the `as_pandas=True`. But for all constraint array with length > 1 this is very helpful. 

A remark on the shadow prices: As all constraints are attached to components, you can specify of which **components** you want to retrieve the shadow_prices. Default is `GlobalConstraint`, `Line` and `Bus`. But you can extend it by 

    n.lopf(pyomo=False, keep_shadowprices=['Bus', 'GlobalConstraint', 'Link', 'Line', 'Generator'])

which will result in retrieving the SP from these components. So for all the constraints which are attached to those components the SP be given stored, under the name one gave the function `set_conref`. 
Sorry, this is a bit jumbled, perhaps better when I make a small notebook with a proper explanation

**Comment by FabianHofmann** [comment]
> Thanks @FabianHofmann! What is
> `lhs += join_exprs(v) + '\n'`

This is the `sum` over all terms (the `+=` is actually a `=`), but there is a better option : 

```
terms = linexpr((n.lines.length[ext_i], get_var(n,  'Line', 's_nom')[ext_i]), as_pandas=True)
lhs = terms.sum()
```

**Comment by nworbmot** [comment]
But what about constraints that aren't attached to components?
I'm a bit concerned that the whole structure is too rigid.

**Comment by FabianHofmann** [comment]
> But what about constraints that aren't attached to components?
> I'm a bit concerned that the whole structure is too rigid.

Mmh, yes so the idea was, that all constraint which are explicitly touching single devices, like fixing capacities per single links, are going to the respective component dataframe or time-series. 
As soon as you have broader constraints like fixing capacities per country, these go into the `GlobalConstraint` component, shadow prices are retrieved automatically.

**Comment by nworbmot** [comment]
But e.g. country-specific constraints will add lot of mess to GlobalConstraints.
Does this mean GlobalConstraints are added outside of the normal network.add() interface?
Often we're also implementing constraints on a subset of components, e.g. to all battery links, or all CHP links, not the whole set of components.
The philosophy of PyPSA has always been to keep it simple and straightforward, and then let the user build complexity on top of that. At the moment I'm not sure that the new code satisfies this philosophy.

**Comment by FabianHofmann** [comment]
okay, fair point. How about the variables? They are also attached to components in the code. If this also applies to them, then I think the whole concept would need a revise and therefore code should be completely restructured, or rather reverted back to the old version (or a mixture between both), also fine if necessary... 

If not, I think it would be doable to bypass the necessity of adding a constraint to a component. Probably creating a Series n.custom_constraints which stores the references would do it. The 'pseudo'-component would then work like this 

```
set_conref(n, con, 'CustomConstraint', 'myconstraint')
```
All shadow prices would then afterwards be in n.custom_constraints

**Comment by FabianHofmann** [comment]
Another idea: 

`set_varref` and `set_conref` can both take components as argument, but also any other names. For example if I want to construct a constraint for all chp links, I could write 

```
con = some_defined_and_written_constraints
set_conref(n, con, 'chp_constraint', 'some_constraintname')
```
Then I would find the constraints in `n.chp_constraint` with column `some_constraintname`. Then the `keep_shadowprices` argument has to include 'chp_constraint' to retrieve the SP.

**Comment by nworbmot** [comment]
That could work. Let's think about it. I think it would help to look at some actual usage examples. I can definitely imagine people adding extra variables that have nothing to do with components (e.g. piecewise learning curves for technology costs). The system should be as flexible as Pyomo itself, otherwise the advantage is not there for customisability.

**Comment by FabianHofmann** [comment]
Hey, @nworbmot,
I had a small coffe and coding session. I managed to tackle some of your concerns (I hope):

* all variables and constraints are no flowing into `n.vars` and `n.cons` under the name one has specified. An info of all variables is in n.variables (or n.constraints).

* two new front end functions `define_variables` and `define_constraints` which do all things needed (writing and storing references) were introduced
For example I can have the extra functionaltity:
  ```
  def extra_functionality(n, snapshots):
      vars_link = get_var(n, 'Link', 'p_nom')
      eff = n.links.at['battery_power', 'efficiency']
      lhs = linexpr((1, vars_link['battery_power']), (-eff, vars_link['battery_discharge']))
      define_constraints(n, lhs, '=', 0, 'battery_discharge', attr='fixratio')
  ``` 
  The resulting constraint will be stored at `n.cons.battery_discharge.df` under column 'fixratio'. If I had defined a time-dependent constraint it would go into `n.cons.battery_discharge.pnl.fixratio`. So there is a differentiation whether it is a static or time-dependent constraint. 

 
* variables and constraints which are not arrays are safely treated (as in the example above), I think this led to an error before, because they had no defined length.

* all duals and solutions which no do have a predefined place to go, are stored into `n.duals` and `n.sols`. So, in the example the dual is stored at `n.duals.battery_discharge.df.fixratio` if
    * 'battery_discharge' is in `keep_shadowprices` 
    * `keep_shadowprices` is True (then all sp all retrieved)

   Custom variables are always getting their solution in `n.sols`

**Comment by FabianHofmann** [comment]
There is a small error, right now in the code, so don't use it right now. Nothing serious, I'll fix later

**Comment by nworbmot** [comment]
I've checked the new code with a complicated extra_functionality use case and it works very nicely: the code is compact and the result agrees with Pyomo.
In general I think the current design is better: the variables and constraints are customisable and kept away from the other data tables (where they can make a mess if the LOPF fails with an error).
I still want to think about the structure next week, but I don't think we need major changes.
We need to document how linexpr handles Series and DataFrames with different numbers of axes. The alignment check in `linopt.broadcasted_axes` at `assert (axes[-1] == df.axes[-1]).all()` only checks the final axis?
Small problems I noticed:
1. The optimal capacities `*_nom_opt` are not being saved (I get all zeroes).
2. The constant in the objective function is not working, i.e. it's not subtracting the capital costs of existing assets defined via `*_nom`.
Maybe these two are related. Strange that unit testing is still working.

**Comment by nworbmot** [comment]
One more suggestion: to save memory, wouldn't it be better to save the variable and constraint references as integers rather than strings? The "x" and "c" can be added later when the constraints and variables are written.

**Comment by FabianHofmann** [comment]
I just finished a [jupyter-notebook](https://github.com/PyPSA/PyPSA/blob/nomopyomo/examples/lopf_with_pyomo_False.ipynb) with examples of extra_functionalities. 

> 1. The optimal capacities *_nom_opt are not being saved (I get all zeroes).

yes, is fixed now (was accidentely stored in "nomopt" without underscore

> 2. The constant in the objective function is not working, i.e. it's not subtracting the capital costs of existing assets defined via *_nom

I'll have a look at that

> 3. The storing of StorageUnit p_dispatch and p_store is breaking imports, since they're not in the standard component_attrs - these need to be updated.

I'm wondering.. there are in the component_attrs ([here](https://github.com/PyPSA/PyPSA/blob/nomopyomo/pypsa/component_attrs/storage_units.csv#L27), but I just tested netcdf export and import...  


Considering the 'x's and and 'c's, I cannot see an option for adding them later. As `linexpr` already brings strings together it is hard to distinguish later what is coefficient and what is variable. 
One idea though: When creating an array of linear expressions, one could always keep coefficient and variable as a tuple, i.e.  a DataFrame or Series with entries like (0.5, 11)  where 0.5 is the coefficient and 11 is what was 'x11' before.

**Comment by FabianHofmann** [comment]
>  The alignment check in linopt.broadcasted_axes at assert (axes[-1] == df.axes[-1]).all() only checks the final axis? 

Yes, it comes from broadcasting operations. Like when you calculate a Series times a DataFrame, the operation 'roles' over all rows of the DataFrame. So only the columns of the DataFrame have to be aligned with the index of the Series. 

But it assumes that when you make a operation with two DataFrames that the rows are in aligned order (meaning the snapshots)

**Comment by coroa** [comment]
> > 1. The constant in the objective function is not working, i.e. it's not subtracting the capital costs of existing assets defined via *_nom
> 
> I'll have a look at that

Hearsay, which may be related: Some solvers do not interpret the constant at all, pyomo works around this by declaring an additional variable, fixing it to 1 and then adding the term `constant * newvar` to the objective, as you can observe, like

```python
m = po.ConcreteModel()                                                 
m.x = po.Var(bounds=(0, None))                                         
m.con = po.Constraint(expr=m.x <= 2)                                  
m.objective = po.Objective(expr=m.x + 1000)                            
m.write("test.lp", io_options={'symbolic_solver_labels': True})
```
gives
```                                                           
\* Source Pyomo model name=unknown *\

min 
objective:
+1 x
+1000 ONE_VAR_CONSTANT

s.t.

c_u_con_:
+1 x
<= 2

c_e_ONE_VAR_CONSTANT: 
ONE_VAR_CONSTANT = 1.0

bounds
   0 <= x <= +inf
end
```

**Comment by FabianHofmann** [comment]
@coroa thanks for the advice! It's very similar to how I did it in the end. Only difference: The dummy constant variable is defined in the bounds section and bounded like 

    1000 <= x1 <= 1000

x1 being the objective constant variable, going into the objective with `+1 x1`. It works for all cbc, gurobi and glpk. Think that should be fine as well...

**Comment by nworbmot** [comment]
Back again! I'd like to make the case for not storing variable names and constraint names with strings like "x124" and "c567". If we're just iterating integers for the names, we should store them as integers. This is less memory intensive and corresponds to how we use them.
It would not require a major change, just changing `expr = expr + _str_array(coeff) + _str_array(var) + '\n'` in the function linexpr at https://github.com/PyPSA/PyPSA/blob/69c9880a760dc7e2f485d541ebea01f9eedfe0d1/pypsa/linopt.py#L331 to `expr = expr + _str_array(coeff) + 'x' + _str_array(var) + '\n'`.
@coroa Any thoughts?

**Comment by FabianHofmann** [comment]
that makes totally sense (didn't think of it before!)

**Comment by FabianHofmann** [comment]
@nworbmot 
[Here](https://github.com/PyPSA/PyPSA/commit/f0e533bc55e4ededbf4595ec2cf7ae7cd3dafc0e) is a possible setup. It work as fast (perhaps a tiiiiny bit slower) but consumes less memory

**Comment by nworbmot** [comment]
Great! I've just integrated the code into the solve_network of PyPSA-Eur-Sec. It mostly works, but pne thing that breaks is the test of `status` and `termination_condition` because pyomo returns `ok` for `status` on success whereas nomopyomo returns `optimal`. Could you make this similar to pyomo?

**Comment by FabianHofmann** [comment]
done :)

**Comment by nworbmot** [comment]
Brilliant, thank you! Just running a test for it now.

**Comment by nworbmot** [comment]
Total memory usage (PyPSA+gurobi) is 10% lower with new integer storing of variable and constraint references! Good result!

**Comment by nworbmot** [comment]
I seem to be having the problem that the results are not saved in the netCDF correctly - could it be that the reading out of results with the new integer scheme is broken?

**Comment by nworbmot** [comment]
BTW My changes to the PyPSA-Eur-Sec solving script (very similar to PyPSA-Eur) you can find here:
https://github.com/nworbmot/pypsa-eur-sec/tree/nomopyomo

**Comment by FabianHofmann** [comment]
> I seem to be having the problem that the results are not saved in the netCDF correctly - could it be that the reading out of results with the new integer scheme is broken?

I'm wondering, could you try the following snippet:
```
import pypsa
import os

network_path = os.path.join(pypsa.__path__[0], '..', 'examples', 'ac-dc-meshed',
                            'ac-dc-data')
n = pypsa.Network(network_path)
n.lopf(pyomo=False)

n.export_to_netcdf('test.nc')
n2 = pypsa.Network('test.nc')
print(n.generators_t.p)
print(n2.generators_t.p)
```

It works fine for me. The only thing that happens is the order of the generators which is sorted alphabetically afterwards and was aligned to `n.generators.index` before.

**Comment by nworbmot** [comment]
Hey, I've solved the problem with the above commit. The problem was with gurobi in `linopt` - the `termination_condition` was being set to `status`, i.e. "ok", which is then rejected as "Problem was not solved to optimality" in `linopf`.
BTW Here's the new memory picture - consumption is less than a third of Pyomo!
pyomo-versus-fabian-191202

**Comment by FabianHofmann** [comment]
> Hey, I've solved the problem with the above commit. The problem was with gurobi in `linopt` - the `termination_condition` was being set to `status`, i.e. "ok", which is then rejected as "Problem was not solved to optimality" in `linopf`.

aah, sorry about that, overlooked it

**Comment by nworbmot** [comment]
No worries. One thing we may have to deal with later, is that some large problems don't solve with "ok", but with "warning", which is why @coroa wrote this exception:
https://github.com/PyPSA/pypsa-eur/blob/master/scripts/solve_network.py#L294
The results are still useable, so we may need a work around. I suggest to fix this when it arises, since I can't remember exactly what the status should be...

**Comment by nworbmot** [comment]
Hey, I'm just finalising the implementation of nomopyomo in PyPSA-Eur. I'm having a problem when adding a line volume constraint to the problem which already has a CO2 constraint. E.g. take the AC-DC example from PyPSA/examples and add before LOPF:

```
n.links["carrier"] = "DC"

n.add("GlobalConstraint",
          "line_volume_constraint",
          type="transmission_volume_expansion_limit",
          carrier_attribute="AC,DC",
          sense="<=",
          constant=1e6)
```
This results in an error:
```
Traceback (most recent call last):                                                                                                                                     File "/mnt/lsdf/tom/playground/pypsa-eur-sec/.snakemake/shadow/tmpl0ik_sbe/.snakemake/scripts/tmpp_9_bh7z.solve_network.py", line 372, in <module>                     n = solve_network(n)                                                                                                                                               File "/mnt/lsdf/tom/playground/pypsa-eur-sec/.snakemake/shadow/tmpl0ik_sbe/.snakemake/scripts/tmpp_9_bh7z.solve_network.py", line 304, in solve_network                status, termination_condition = run_lopf(n, allow_warning_status=True)                                                                                             File "/mnt/lsdf/tom/playground/pypsa-eur-sec/.snakemake/shadow/tmpl0ik_sbe/.snakemake/scripts/tmpp_9_bh7z.solve_network.py", line 253, in run_lopf                     formulation=solve_opts['formulation'])                                                                                                                             File "pypsa/pypsa/components.py", line 502, in lopf                                                                                                                    return network_lopf_lowmem(self, **args)                                                                                                                           File "pypsa/pypsa/linopf.py", line 882, in network_lopf                                                                                                                extra_functionality, solver_dir)                                                                                                                                   File "pypsa/pypsa/linopf.py", line 622, in prepare_lopf                                                                                                                define_global_constraints(n, snapshots)                                                                                                                            File "pypsa/pypsa/linopf.py", line 512, in define_global_constraints                                                                                                   set_conref(n, con, 'GlobalConstraint', 'mu', name)                                                                                                                 File "pypsa/pypsa/linopt.py", line 429, in set_conref                                                                                                                  _add_reference(n.cons[c], constraints, attr, pnl=pnl)                                                                                                              File "pypsa/pypsa/linopt.py", line 387, in _add_reference                                                                                                              ref_dict.df.loc[df.index, attr] = df                                                                                                                               File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 204, in __setitem__                                                         indexer = self._get_setitem_indexer(key)                                                                                                                           File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 182, in _get_setitem_indexer                                                return self._convert_tuple(key, is_setter=True)                                                                                                                    File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 262, in _convert_tuple                                                      idx = self._convert_to_indexer(k, axis=i, is_setter=is_setter)                                                                                                     File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1285, in _convert_to_indexer                                                return self._get_listlike_indexer(obj, axis, **kwargs)[1]                                                                                                          File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1092, in _get_listlike_indexer                                              keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing                                                                                             File "/opt/conda/envs/dockerenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1177, in _validate_read_indexer                                             key=key, axis=self.obj._get_axis_name(axis)                                                                                                                      KeyError: "None of [Index(['line_volume_constraint'], dtype='object')] are in the [index]" 
```
Any ideas?

**Comment by FabianHofmann** [comment]
Could you try setting 

    ref_dict.df = pd.concat([ref_dict.df, df.to_frame(attr)])

instead where the error happens (387 in linopt)? I think it worked for me

**Comment by nworbmot** [comment]
That works for the AC-DC example! I'll try it on PyPSA-Eur now.

**Comment by nworbmot** [comment]
There was no error building problem for PyPSA-Eur now. Just waiting for it to solve. Did the error make sense, i.e. did you understand why it was going wrong?

**Comment by FabianHofmann** [comment]
> There was no error building problem for PyPSA-Eur now. Just waiting for it to solve. Did the error make sense, i.e. did you understand why it was going wrong?

I'm not very sure why specifically this works now. Before it had to do with changing adding multiple values at the same time, like

```python 
df = pd.DataFrame(1, index=['a'], columns=[0])
df.loc[['b']] = 2
```
which breaks (I did not know this before). Now, what happens is rewriting the whole dataframe which is more stable.

**Comment by nworbmot** [comment]
PyPSA-Eur-Sec is now all working for arbitrary line volumes:
https://github.com/nworbmot/pypsa-eur-sec/tree/nomopyomo
I think we can aim for a new PyPSA release with it before Xmas? Anything else that needs sorting before we release?
Thank you @FabianHofmann for all your hard working integrating it!

**Comment by FabianHofmann** [comment]
@nworbmot this sounds good! Perhaps to little things: 

1. Did you test a unit commitment model? the implementation allows only a light version of UC but it should work, a warning is given which parameters are ignored.

2. There is the new module `stats.py` which is supposed to enable double-checking the constraints post-optimization. Should we include it? Or any changes to made here?  We could also merge it later...

**Comment by nworbmot** [comment]
Thanks @FabianHofmann for all your hard work on this! By far the biggest pull request we've ever had!

**Pull request merged**